pkg/logs/input/docker/launcher.go

```diff
@@ -0,0 +1,145 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build docker
+
+package docker
+
+import (
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/tagger"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+
+	"github.com/DataDog/datadog-agent/pkg/logs/auditor"
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
+	"github.com/DataDog/datadog-agent/pkg/logs/pipeline"
+	"github.com/DataDog/datadog-agent/pkg/logs/restart"
+)
+
+const scanPeriod = 10 * time.Second
+
+// A Launcher listens for stdout and stderr of containers
+type Launcher struct {
+	pipelineProvider pipeline.Provider
+	sources          *config.LogSources
+	tailers          map[string]*Tailer
+	cli              *client.Client
+	registry         auditor.Registry
+	stop             chan struct{}
+}
+
+// NewLauncher returns an initialized Launcher
+func NewLauncher(sources *config.LogSources, pipelineProvider pipeline.Provider, registry auditor.Registry) (*Launcher, error) {
+	scanner := &Launcher{
+		pipelineProvider: pipelineProvider,
+		sources:          sources,
+		tailers:          make(map[string]*Tailer),
+		registry:         registry,
+		stop:             make(chan struct{}),
+	}
+	err := scanner.setup()
+	if err != nil {
+		return nil, err
+	}
+	return scanner, nil
+}
+
+// setup initializes the docker client and the tagger,
+// returns an error if it failed.
+func (l *Launcher) setup() error {
+	var err error
+	// create a new docker client
+	l.cli, err = NewClient()
+	if err != nil {
+		return err
+	}
+	// initialize the tagger
+	err = tagger.Init()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Start starts the Launcher
+func (l *Launcher) Start() {
+	go l.run()
+}
+
+// Stop stops the Launcher and its tailers in parallel,
+// this call returns only when all the tailers are stopped
+func (l *Launcher) Stop() {
+	l.stop <- struct{}{}
+	stopper := restart.NewParallelStopper()
+	for _, tailer := range l.tailers {
+		stopper.Add(tailer)
+		delete(l.tailers, tailer.ContainerID)
+	}
+	stopper.Stop()
+}
+
+// run checks periodically which docker containers are running until stop
+func (l *Launcher) run() {
+	for {
+		select {
+		case source := <-l.sources.GetSourceStreamForType(config.DockerType):
+			l.launch(source)
+		case <-l.stop:
+			// no docker container should be tailed anymore
+			return
+		}
+	}
+}
+
+// launch launches a new tailer for the source.
+func (l *Launcher) launch(source *config.LogSource) {
+	// containerID := source.EntityID
+	containerID := ""
+	if _, isTailed := l.tailers[containerID]; isTailed {
```

Comment:
I'd love to see more comments explaining the behaviour we expect from this function.
It's even more important as what we get from `l.sources.GetSourceStreamForType(config.DockerType)` can be considered as a blackbox with a proper api
For instance, in which case would we already be tailing the container?